[ELY-2078] Add encryption and integrity support to FileSystemSecurityRealm

Ashpan opened the pull request:
https://issues.redhat.com/browse/ELY-2078

Ashpan committed:
[ELY-2078] Add encryption and integrity support to FileSystemSecurityRealm

Ashpan committed:
[ELY-2078] Elytron tool to encrypted a filesystem realm

Skyllarr commented on tests/base/src/test/java/org/wildfly/security/auth/FileSystemSecurityRealmTest.java:
```diff
@@ -552,9 +920,59 @@ public void testPartialIteratingTryWithResource() throws Exception {
+                .setSecretKey(SecretKeyUtil.generateSecretKey(192))
+                .build();
+        ModifiableRealmIdentity existingIdentity = securityRealm.getRealmIdentityForUpdate(new NamePrincipal("plainUser"));
+        existingIdentity.verifyEvidence(new PasswordGuessEvidence(actualPassword));
```
@Ashpan With different key it is still possible to assess that the identity exists, since the username is base32 encoded, and only the reading of its attributes and the credential verification fails.

Skyllarr replied:
I read your analysis document and the limitation of the username and why this was decided is well explained, thank you. I was thinking whether the existence of the identity can be obscured somehow in these cases, but didn't come up with anything that would help when the store is exposed. So this makes sense.

Skyllarr commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -205,9 +224,11 @@ public FileSystemSecurityRealm(final Path root, final int levels) {
     public FileSystemSecurityRealm(final Path root, final int levels, final Encoding hashEncoding, final Charset hashCharset) {
-        this(root, NameRewriter.IDENTITY_REWRITER, levels, true, hashEncoding, hashCharset);
+        this(root, NameRewriter.IDENTITY_REWRITER, levels, true, hashEncoding, hashCharset
+                , null);
```
@Ashpan  just a minor, there are some unnecessary newlines and padding in this file. Is there a reason why this is still marked as a Draft?

fjuma reviewed (changes requested):
Thanks, @Ashpan! I've added some comments.

Ashpan committed:
[ELY-2078] Verify Filesystem Integrity

darranl committed:
[ELY-2241] Bump to a new version 1.1 namespace for the fileysystem realm.

This new schema drops all elements and types specific to authentication
client.

Additionally prepare for support for new items to be added to this
namespace.

rmartinc committed:
[ELY-2294] Failures in the WFCORE test-suite with branch 1.x

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -306,14 +298,10 @@ private Path pathFor(String name) {
-
-        if (encoded) {
-            normalizedName = normalizedName + "-" + base32;
+        if (secretKey != null | encoded) {
```
Sorry, I had a typo in my original comment for this one. We should use `||` here (instead of just `|`).

Ashpan replied:
Fixed

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/ElytronMessages.java:
```diff
@@ -146,4 +146,11 @@
     @Message(id = 13004, value = "JAAS logout failed for principal %s")
     void debugInfoJaasLogoutFailure(Principal principal, @Cause Throwable cause);
+
+    @Message(id = 13005, value = "Signature for the following identity is invalid: %s.")
```
Very minor comment, the trailing period can be removed here and below.

Ashpan replied:
Removed

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -135,22 +218,44 @@
      * @param hashCharset the character set to use when converting password strings to a byte array. Uses UTF-8 by default.
      * @param hashEncoding the string format for the hashed passwords. Uses Base64 by default.
+     * @param secretKey the SecretKey used to encrypt and decrypt the security realm (if {@code null}, the security realm will be unencrypted)
+     * @param providers The providers supplier
```
privateKey and publicKey should be added here too.

Just minor, but secretKey should also be moved under providers.

Ashpan replied:
Added and moved

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -135,22 +218,46 @@
+     * @param publicKey the PublicKey used for verifying the signature of the identities in security realm (if {@code null}, the security realm will not have integrity enabled)
      */
-    public FileSystemSecurityRealm(final Path root, final NameRewriter nameRewriter, final int levels, final boolean encoded, final Encoding hashEncoding, final Charset hashCharset, Supplier<Provider[]> providers) {
+    public FileSystemSecurityRealm(final Path root, final NameRewriter nameRewriter, final int levels, final boolean encoded, final Encoding hashEncoding, final Charset hashCharset, Supplier<Provider[]> providers, final SecretKey secretKey, final PrivateKey privateKey, final PublicKey publicKey) {
```
Just thinking about this one some more. Since the intent going forwards is for users to make use of the builder to create a FileSystemSecurityRealm, perhaps "public" could be removed from this constructor?

Ashpan replied:
Makes sense to me. Removed `public`

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -135,22 +218,46 @@
+        this.providers = providers != null ? providers : INSTALLED_PROVIDERS;
+        this.secretKey = secretKey;
+        this.privateKey = publicKey != null ? privateKey : null;
+        this.publicKey = privateKey != null ? publicKey : null;
```
Once the constructor is package private, it should only ever be called with valid params (since only the build method and the other constructors can call it). So you wouldn't need these checks here.

Ashpan replied:
Moved any checks into the FileSystemSecurityRealmBuilder

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -599,6 +720,16 @@ private Path tempPath() {
+                    }
+                    if ( this.publicKey != null && ! validateDigitalSignature(this.path.toString(), this.name)) {
+                        throw ElytronMessages.log.invalidIdentitySignature(this.name);
+                    }
```
It looks like the above lines first create the digital signature then verify it immediately. I don't think the immediate verification is needed there.

Ashpan replied:
Good catch, removed

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -745,6 +879,10 @@ private Void replaceIdentityPrivileged(final LoadedIdentity newIdentity) throws
                         // nothing we can do
                     }
+                    if ( this.publicKey != null ) { createDigitalSignature(this.path.toString(), this.name); }
+                    if ( this.publicKey != null && ! validateDigitalSignature(this.path.toString(), this.name)) {
```
Same here with respect to immediate verification.

Ashpan replied:
Removed

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -757,12 +895,26 @@ private Void replaceIdentityPrivileged(final LoadedIdentity newIdentity) throws
+            // to an older version.
+
+            // Do we require version 1.1?
+            // return Version.VERSION_1_1;
```
Looks like we'd need to require the new version if integrity is enabled (i.e., if publicKey, privateKey aren't null).

Ashpan replied:
Changed

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -903,31 +1069,32 @@ private LoadedIdentity parseIdentityContents(final XMLStreamReader streamReader,
+                    return new LoadedIdentity(name, credentials, attributes, hashEncoding, providers);
                 }
-                if (! validNamespace(streamReader.getNamespaceURI())) {
+                if ( !(version.getNamespace().equals(streamReader.getNamespaceURI())) && !(XMLSignature.XMLNS.equals(streamReader.getNamespaceURI())) ) {
```
Have you tried parsing an older version of an identity file to make sure that still works?

Ashpan replied:
Yeah, it still works with this check

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -1160,22 +1352,95 @@ private void consumeContent(final XMLStreamReader reader) throws XMLStreamExcept
+                Document doc = dbf.newDocumentBuilder().parse(path);
+                NodeList nl = doc.getElementsByTagNameNS(XMLSignature.XMLNS, "Signature");
+                if (nl.getLength() == 0) {
+                    throw new RealmUnavailableException("Cannot find Signature element");
```
This message should be in ElytronMessages.

Ashpan replied:
Moved to ElytronMessages

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -1160,22 +1352,95 @@ private void consumeContent(final XMLStreamReader reader) throws XMLStreamExcept
+                DOMValidateContext valContext = new DOMValidateContext(publicKey, nl.item(0));
+                XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+                boolean coreValidity = signature.validate(valContext);
+                ElytronMessages.log.tracef("FileSystemSecurityRealm - verification against signature for credential [%s] = %b", name, coreValidity);
```
Is `name` the username? If so, maybe the text should be changed from "credential" to "user"

Ashpan replied:
Yes it is, changed the text to `verification against signature for user [%s]`

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -1160,22 +1352,95 @@ private void consumeContent(final XMLStreamReader reader) throws XMLStreamExcept
+                boolean coreValidity = signature.validate(valContext);
+                ElytronMessages.log.tracef("FileSystemSecurityRealm - verification against signature for credential [%s] = %b", name, coreValidity);
+                return coreValidity;
+           } catch (ParserConfigurationException | IOException | MarshalException | XMLSignatureException | org.xml.sax.SAXException e) {
```
Just to check, is there a reason for using `org.xml.sax.SAXException` explicitly?

Ashpan replied:
I missed that when adding the imports, importing it instead now

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealmBuilder.java:
```diff
@@ -0,0 +1,200 @@
+        }
+
+        if (privateKey == null ^ publicKey == null) {
+            throw new IllegalArgumentException("Both PrivateKey and PublicKey must be defined for realm at: " + root);
```
Should be in ElytronMessages.

Ashpan replied:
Moved to ElytronMessages

fjuma commented on tests/base/src/test/java/org/wildfly/security/auth/FileSystemSecurityRealmTest.java:
```diff
@@ -79,6 +92,26 @@
+    KeyPairGenerator keyPairGen = KeyPairGenerator.getInstance("RSA");
+    KeyPair pair = keyPairGen.generateKeyPair();
+    PrivateKey privateKey = pair.getPrivate();
+    PublicKey publicKey = pair.getPublic();
```
Very minor but these variable declarations could moved up under the provider declaration for better readability.

fjuma replied:
They can also be private static final.

Ashpan replied:
Moved up, and changed to private static. Couldn't make it final since a try catch was required around the KeyPair definition and SecretKey definition

fjuma commented on tests/base/src/test/java/org/wildfly/security/auth/FileSystemSecurityRealmTest.java:
```diff
@@ -116,10 +167,67 @@ public void testCreateAndLoadIdentity() throws Exception {
+                .setPublicKey(publicKey)
+                .setPrivateKey(privateKey)
+                .build();
+        ModifiableRealmIdentity existingIdentity = securityRealm.getRealmIdentityForUpdate(new NamePrincipal("user"));
```
It looks like this test is relying on a hardcoded identity file with an invalid signature? If so, a comment should be added here to indicate that.

Ashpan replied:
Yes it is, comment added

fjuma commented on tests/base/src/test/java/org/wildfly/security/auth/FileSystemSecurityRealmTest.java:
```diff
@@ -116,10 +167,67 @@ public void testCreateAndLoadIdentity() throws Exception {
+        char[] actualPassword = "secretPassword".toCharArray();
+        existingIdentity.verifyEvidence(new PasswordGuessEvidence(actualPassword));
+        existingIdentity.dispose();
+    }
```
If you haven't already added it, would be good to have a test that calls build with invalid params (e.g., missing private key).

Ashpan replied:
Added `testInvalidBuild()`

fjuma commented on tests/base/src/test/java/org/wildfly/security/auth/FileSystemSecurityRealmTest.java:
```diff
@@ -537,18 +1070,146 @@ public void testPartialIteratingTryWithResource() throws Exception {
     }
 
+    @Test(expected = RealmUnavailableException.class)
+    public void testMismatchSecretKey() throws Exception {
```
Do you have a similar test for integrity (i.e., wrong private key / public key)

Ashpan replied:
Just added `testMismatchPublicKey()` and `testMismatchPrivateKey()`

fjuma commented on tests/base/src/test/resources/filesystem-realm/u/s/e/user-OVZWK4Q.xml:
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?><identity xmlns="urn:elytron:1.0">
```
"urn:elytron:1.0" --> looks like the schema version here should be updated

fjuma commented on tool/src/main/java/org/wildfly/security/tool/FileSystemEncryptRealmCommand.java:
```diff
@@ -0,0 +1,770 @@
+                try {
+                    descriptor.setHashCharset(Charset.forName(hashCharsetOption));
+                } catch (UnsupportedCharsetException e) {
+//                    Error
```
Looks like error handling needs to be added here and below (there should be some examples in the existing commands).

Ashpan replied:
Added

fjuma commented on tool/src/main/java/org/wildfly/security/tool/FileSystemEncryptRealmCommand.java:
```diff
@@ -0,0 +1,770 @@
+//                    Error
+                }
+            }
+                if (credentialStoreOption != null && secretKeyAliasOption != null) {
```
Looks like there's extra spaces at the beginning of the if here.

Ashpan replied:
Fixed

fjuma commented on tool/src/main/java/org/wildfly/security/tool/FileSystemEncryptRealmCommand.java:
```diff
@@ -0,0 +1,770 @@
+    private void findMissingRequiredValuesAndSetValues(int count, Descriptor descriptor) {
+        boolean missingRequiredValue = false;
+        if (descriptor.getInputRealmLocation() == null) {
+            warningHandler(ElytronToolMessages.msg.skippingDescriptorBlock(count, "missing input realm location"));
```
It would be better to incorporate the reason into the ElytronToolMessages so all the text is there.

Ashpan replied:
Added

fjuma commented on tool/src/test/java/org/wildfly/security/tool/FileSystemEncryptRealmCommandTest.java:
```diff
@@ -0,0 +1,139 @@
+            if(!fileExists("target/test-classes/filesystem-encrypt/fs-encrypted-realms/"+file)){
+                throw new FileNotFoundException("Missing file from Bulk Descriptor File: " + file);
+            } else {
+                System.out.println("Found file: " + file);
```
I think this print statement and the ones below can be removed.

Ashpan replied:
Removed

fjuma commented on auth/realm/base/src/main/resources/schema/elytron-identity-1_1.xsd:
```diff
@@ -0,0 +1,147 @@
+
+<xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema"
+            targetNamespace="urn:elytron:identity:1.0.1"
+            xmlns="urn:elytron:identity:1.0.1"
```
Shouldn't this be 1.1 (to match elytron-identity-1_1.xsd?)

Ashpan committed:
Added identity file hashing and main index

It will encrypt main_index, if encryption is enabled

skyllarr commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemRealmUtil.java:
```diff
@@ -0,0 +1,52 @@
+            ModifiableRealmIdentity newIdentity = encryptedRealm.getRealmIdentityForUpdate(new NamePrincipal(identity.getRealmIdentityPrincipal().getName()));
+            newIdentity.create();
+            newIdentity.setCredentials(credentials);
+            newIdentity.setAttributes(attributes);
```
minor since this is used by tool only but newIdentity.dispose() can be added at the end

Ashpan replied:
Added

skyllarr commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -113,20 +154,67 @@
+            return namespace;
+        }
+
+        boolean isAtLeast(Version version) {
```
unused method

Ashpan replied:
This was added by Darran for a namespace bump for identity files.
It's unused right now but I believe it's for the future
https://github.com/wildfly-security/wildfly-elytron/pull/1613/files#diff-ae7c68238e8237a3c02a3c53f9204a3981c65a315ca7959878681276be0bf3f7R144-R146

skyllarr commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -885,9 +1040,17 @@ private LoadedIdentity loadIdentityPrivileged(final boolean skipCredentials, fin
-                    return parseIdentity(streamReader, skipCredentials, skipAttributes);
+                    LoadedIdentity parsedIdentity = parseIdentity(streamReader, skipCredentials, skipAttributes);
+                    if (this.publicKey != null) {
+                        createDigitalSignature(this.path.toString(), this.name);
```
can createDigitalSignature be deleted from this method?

Ashpan replied:
Ah yes it can, will remove

skyllarr commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java:
```diff
@@ -1177,22 +1371,220 @@ private void consumeContent(final XMLStreamReader reader) throws XMLStreamExcept
+                    throw ElytronMessages.log.unableToAccessMainIndex(this.mainIndex.toString());
+                }
+                if (!checksumValidity) {
+                    if (!validateDigitalSignature(this.path.toString(), this.name)) {
```
should this be `if (!checksumValidity || !validateDigitalSignature())` ?

Ashpan replied:
If the intention here is the only run the `validateDigitalSignature()` if the `checksumValidity()` is false, is this the correct way to go?

skyllarr replied:
@Ashpan sorry if I am asking something obvious. Can checksumValidity be false and validateDigitalSignature() return true? Is integrity valid in such a case?

Ashpan replied:
Ah yes so it is possible. This would be the case if some attributes of the file changes but without really changing the contents of the file, like white spaces or something similar. That would make the checksumValidity false, which validateDigitalSignature remains true. Integrity is still valid in this scenario.

skyllarr commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealmBuilder.java:
```diff
@@ -0,0 +1,225 @@
+     * @return this builder.
+     */
+    public FileSystemSecurityRealmBuilder setLevels(final int levels) {
+        this.levels = levels;
```
we can use Assert.checkMinimumParameter("levels", 0, levels)

Ashpan replied:
Added

skyllarr commented on tests/base/src/test/java/org/wildfly/security/auth/FileSystemSecurityRealmTest.java:
```diff
@@ -153,28 +333,169 @@ public void testCaseSensitive() throws Exception {
+                .setLevels(3)
+                .setProviders(ELYTRON_PASSWORD_PROVIDERS);
+        if (mode.equals("encryption")) { securityRealmBuilder.setSecretKey(secretKey); }
+        else if (mode.equals("integrity")) { securityRealmBuilder.setPrivateKey(privateKey); securityRealmBuilder.setPublicKey(publicKey); }
```
is there a test for both encryption and integrity together?

Ashpan replied:
Added `testCreateIdentityWithEverythingEncryptionAndIntegrity()`

Ashpan commented:
Closed this PR and reopened new one with only Encryption changes. 

https://github.com/wildfly-security/wildfly-elytron/pull/1676

fjuma commented on tool/src/main/java/org/wildfly/security/tool/FileSystemEncryptRealmCommand.java:
```diff
@@ -0,0 +1,742 @@
+            String secretKeyAlias = descriptor.getSecretKeyAlias();
+            int levels = descriptor.getLevels();
+            Encoding hashEncoding = descriptor.getHashEncoding();
+            Charset hashCharset = descriptor.getHashCharset();
```
Is this needed?
